es6/animations.js

```diff
@@ -5,180 +5,176 @@ import {
-        if (end_state === undefined)
-            end_state = start_state + amt
-        if (amt === undefined)
-            amt = end_state - start_state
-
-        if (start_state + amt != end_state) {
-            console.log({start_state, end_state, amt})
-            throw 'Conflicting values, Animation end_state != start + amt'
+export const computeTheOther = (start, delta, end) => {
+    if (typeof(start) === 'object') {
+        let new_delta = delta ? {...delta} : {}
+        let new_end = end ? {...end} : {}
+        Object.keys(start).forEach((key) => {
+            let [_delta, _end] = computeTheOther(start[key],
+                                                 new_delta[key],
+                                                 new_end[key])
+            new_delta[key] = _delta
+            new_end[key] = _end
+        })
+        return [new_delta, new_end]
+    }
+    if (typeof(start) === 'number') {
+        if (end === undefined && delta !== undefined) {
+            return [delta, start + delta]
+        } else if (end !== undefined && delta === undefined) {
+            return [end - start, end]
+        } else if (start + delta !== end) {
+            console.log((new Error()).stack)
+            throw `start (${start}) + delta (${delta}) !== end (${end})`
         }
+    }
+    return [delta, end]
+}
 
-        return {start_state, end_state, amt}
-    } else {
-        if (start_state === undefined) start_state = {}
-        if (end_state === undefined) end_state = {}
-        if (amt === undefined) amt = {}
+const exactlyOneIsUndefined = (val1, val2) => {
+    return ((val1 === undefined || val2 === undefined)
+            && !(val1 === undefined && val2 === undefined))
+}
 
-        if (key !== undefined)
-            return checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
+const isNumber = (val) => {
+    return typeof(val) === 'number'
+}
 
-        if (typeof(start_state) !== 'object' || typeof(end_state) !== 'object' || typeof(amt) !== 'object') {
-            throw 'Incompatible types passed as {start_state, end_state, amt}, must all be objects or numbers'
-        }
+const validateAnimation = (animation) => {
+    const end_time = animation.end_time
+    const end_state = animation.end_state
 
-        let keys = Object.keys(start_state)
-        if (!keys.length) keys = Object.keys(end_state)
-        if (!keys.length) keys = Object.keys(amt)
-
-        return keys.reduce((acc, key) => {
-            const amts = checked_animation_amt({start_state: start_state[key], end_state: end_state[key], amt: amt[key]})
-            acc.start_state[key] = amts.start_state
-            acc.end_state[key] = amts.end_state
-            acc.amt[key] = amts.amt
-            return acc
-        }, {start_state: {}, end_state: {}, amt: {}})
+    const computed_end_state = computeAnimatedState([animation], _end_time)
+    if (!isEqual(computed_end_state, end_state)) {
```

Comment:
This branch is impossible to reach if `computeTheOther` does its job correctly.

Reply:
There are cases where this isn't true. Custom tick function, e.g.